tests/core/test_signal.py

```diff
@@ -50,6 +52,29 @@ async def test_signals_equality_raises():
         s1 > 4
 
 
+async def test_signal_connect_fails_with_different_backend_on_connection():
+    sim_signal = Signal(MockSignalBackend(str))
+
+    with pytest.raises(ValueError):
+        await sim_signal.connect(mock=True, backend=MockSignalBackend(int))
+
+    with pytest.raises(ValueError):
+        await sim_signal.connect(mock=True, backend=SoftSignalBackend(str))
+
+    with pytest.raises(ValueError):
+        await sim_signal.connect(mock=False, backend=MockSignalBackend(str))
+
+
+async def test_signal_connect_fails_if_different_backend_but_same_by_value():
+    initial_backend = MockSignalBackend(str)
+    sim_signal = Signal(initial_backend)
+
+    with pytest.raises(ValueError):
+        await sim_signal.connect(mock=True, backend=MockSignalBackend(str))
+
+    await sim_signal.connect(mock=True, backend=initial_backend)
```

Comment:
We should skip the `mock` arg in this since we're doing it here, also assert the error string is what we expect.

```suggestion
    with pytest.raises(ValueError) as exc:
        await sim_signal.connect(mock=False, backend=MockSignalBackend(str))
    assert str(exc.value) == "<whatever output string>" 

    await sim_signal.connect(mock=False, backend=initial_backend)
```